datafusion/sql/src/planner.rs

```diff
@@ -2671,6 +2694,49 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             Ok(lit(ScalarValue::new_list(Some(values), data_type)))
         }
     }
+
+    /// Parse number in sql string, convert to Expr::Literal
+    fn parse_sql_number(&self, n: &str) -> Result<Expr> {
+        if let Some(_) = n.find('E') {
+            // not implemented yet
+            // https://github.com/apache/arrow-datafusion/issues/3448
+            Err(DataFusionError::NotImplemented(
+                "sql numeric literals in scientific notation are not supported"
+                    .to_string(),
+            ))
+        } else if let Ok(n) = n.parse::<i64>() {
+            Ok(lit(n))
+        } else if self.options.parse_float_as_decimal {
+            if let Some(i) = n.find('.') {
+                let p = n.len() - 1;
+                let s = n.len() - i - 1;
+                let str = n.replace('.', "");
+                let n = str.parse::<i128>().map_err(|_| {
+                    DataFusionError::from(ParserError(format!(
+                        "Cannot parse {} as i128 when building decimal",
+                        str
```

Comment:
Maybe good to print out original string too, i.e.,

```suggestion
                        "Cannot parse {} as i128 when building decimal. Input token: {}",
                        str, n
```